price-discovery/src/create_pool.rs

```diff
@@ -0,0 +1,103 @@
+elrond_wasm::imports!();
+
+const ACCEPT_FUNDS_FUNC_NAME: &[u8] = b"accept_funds_func";
+
+mod liquidity_pool_proxy {
+    elrond_wasm::imports!();
+
+    #[elrond_wasm::proxy]
+    pub trait LiquidityPool {
+        #[payable("*")]
+        #[endpoint(addInitialLiquidity)]
+        fn add_initial_liquidity(
+            &self,
+            #[payment_multi] payments: ManagedVec<EsdtTokenPayment<Self::Api>>,
+            #[var_args] opt_accept_funds_func: OptionalArg<ManagedBuffer>,
+        ) -> MultiResult3<
+            EsdtTokenPayment<Self::Api>,
+            EsdtTokenPayment<Self::Api>,
+            EsdtTokenPayment<Self::Api>,
+        >;
+    }
+}
+
+#[elrond_wasm::module]
+pub trait CreatePoolModule: crate::common_storage::CommonStorageModule {
+    #[endpoint(createDexLiquidityPool)]
+    fn create_dex_liquidity_pool(&self) -> SCResult<()> {
+        require!(self.lp_token_id().is_empty(), "Pool already created");
+        self.require_deposit_period_ended()?;
+
+        let launched_token_id = self.launched_token_id().get();
+        let accepted_token_id = self.accepted_token_id().get();
+
+        let launched_token_balance = self.blockchain().get_sc_balance(&launched_token_id, 0);
+        let accepted_token_balance = self.blockchain().get_sc_balance(&accepted_token_id, 0);
+
+        require!(
+            launched_token_balance > 0,
+            "No Launched tokens were deposited"
+        );
+        require!(accepted_token_balance > 0, "No users deposited tokens");
+
+        self.launched_token_final_amount()
+            .set(&launched_token_balance);
+        self.accepted_token_final_amount()
+            .set(&accepted_token_balance);
+
+        let mut payments = ManagedVec::<Self::Api, EsdtTokenPayment<Self::Api>>::new();
+        payments.push(EsdtTokenPayment {
+            token_type: EsdtTokenType::Fungible,
+            token_identifier: launched_token_id,
+            token_nonce: 0,
+            amount: launched_token_balance,
+        });
+        payments.push(EsdtTokenPayment {
+            token_type: EsdtTokenType::Fungible,
+            token_identifier: accepted_token_id,
+            token_nonce: 0,
+            amount: accepted_token_balance,
+        });
+
+        let dex_sc_address = self.dex_sc_address().get();
+        let contract_call = self
+            .dex_proxy(dex_sc_address)
+            .add_initial_liquidity(payments, OptionalArg::Some(ACCEPT_FUNDS_FUNC_NAME.into()));
+
+        let (lp_token, _, _) = contract_call.execute_on_dest_context().into_tuple();
+        self.lp_token_id().set(&lp_token.token_identifier);
+        self.total_lp_tokens_received().set(&lp_token.amount);
+
+        Ok(())
+    }
+
+    #[payable("*")]
+    #[endpoint]
+    fn accept_funds_func(&self) -> SCResult<()> {
```

Comment:
maybe add a check that this endpoint is called only once.

Reply:
This is basically just a callback for the dex `add_initial_liquidity`, so it can't be called directly. Additionally, it does not alter the state, so even if someone managed to somehow trick the DEX contracts to call this endpoint, it wouldn't affect anything.